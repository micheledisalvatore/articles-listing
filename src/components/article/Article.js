import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

import { showArticle } from '../../actions/articles';

import { Content, Img, TextWrapper } from './Article.styled';

export const Article = ({ artwork, content, full, id, title, showArticleAction }) => {
  const handleArticleClick = () => { showArticleAction(id) };

  return (
    <div>
      <Img src={artwork} alt={title} />
      <TextWrapper>
        <h3>{title}</h3>
        <Content dangerouslySetInnerHTML={{__html: content}} onClick={handleArticleClick} full={full} />
      </TextWrapper>
    </div>
  )
}

Article.propTypes = {
  artwork: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  full: PropTypes.bool,
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  showArticleAction: PropTypes.func.isRequired,
}

Article.defaultProps = {
  full: false,
}

export const mapDispatchToProps = (dispatch) => ({
  showArticleAction: bindActionCreators(showArticle, dispatch),
})