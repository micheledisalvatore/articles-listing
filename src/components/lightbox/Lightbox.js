import React, { Component } from 'react';
import { bindActionCreators } from 'redux';

import { hideArticle } from '../../actions/articles';

import Article from '../article';

import { Wrapper, Background } from './Lightbox.styled';

export class Lightbox extends Component {
  constructor(props){
    super(props);
    this.handleEscKey = this.handleEscKey.bind(this);
  }

  handleEscKey(event){
    const { hideArticleAction } = this.props;

    if(event.keyCode === 27) {
      hideArticleAction();
    }
  }

  componentDidMount(){
    document.addEventListener("keydown", this.handleEscKey, false);
  }

  componentWillUnmount(){
    document.removeEventListener("keydown", this.handleEscKey, false);
  }

  render() {
    const { hideArticleAction, article: {artwork, content, title }} = this.props;

    return (
      <Wrapper>
        <Background onClick={hideArticleAction} />
        <Article artwork={artwork} content={content} title={title} full />
      </Wrapper>
    )
  }
}

export const mapStateToProps = ({ articles: { list, selected } }) => ({
  article: list[selected],
});

export const mapDispatchToProps = (dispatch) => ({
  hideArticleAction: bindActionCreators(hideArticle, dispatch),
})
