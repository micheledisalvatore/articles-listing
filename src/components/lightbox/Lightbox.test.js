/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';

import Article from '../article';

import { Lightbox, mapStateToProps } from './Lightbox';
import { Background } from './Lightbox.styled';

describe('Given a Lightbox component', () => {
  let component;
  let wrapper;
  let hideArticleActionMock;
  const events = [];

  describe('when it is rendered', () => {
    beforeEach(() => {
      hideArticleActionMock = jest.fn();
      document.addEventListener = jest.fn((event, cb) => {
        events[event] = cb;
      });
      const requiredProps = {
        hideArticleAction: hideArticleActionMock,
        article: {
          artwork: 'http://foo.bar',
          content: 'foo',
          title: 'Bar Title',
        },
      };

      component = <Lightbox {...requiredProps} />;
      wrapper = shallow(component);
    });

    afterEach(() => {
      document.addEventListener.mockReset();
      hideArticleActionMock.mockReset();
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should show the Article in full', () => {
      expect(wrapper.find(Article)).toHaveProp('full', true);
    });

    describe('and the user clicks on the Content', () => {
      beforeEach(() => {
        wrapper.find(Background).simulate('click');
      });

      it('should invoke hideArticleAction', () => {
        expect(hideArticleActionMock).toHaveBeenCalledWith();
      });
    });

    describe('and the user clicks on the Content', () => {
      beforeEach(() => {
        events.keydown({ keyCode: 27 });
      });

      it('should invoke hideArticleAction', () => {
        expect(hideArticleActionMock).toHaveBeenCalledWith();
      });
    });
  });
});

describe('Given a mapStateToProps function', () => {
  describe('when it is called', () => {
    it('should return an object with the passed values', () => {
      expect(mapStateToProps({
        articles: {
          list: {
            1: {
              id: 1,
              name: 'foo',
            },
            2: {
              id: 2,
              name: 'bar',
            },
          },
          selected: 1,
        },
      })).toEqual({
        article: {
          id: 1,
          name: 'foo',
        },
      });
    });
  });
});
