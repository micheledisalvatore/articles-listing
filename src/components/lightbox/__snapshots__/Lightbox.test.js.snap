// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Lightbox component when it is rendered should match the snapshot 1`] = `
<styled.div>
  <styled.button
    onClick={[MockFunction]}
  />
  <Connect(Article)
    artwork="http://foo.bar"
    content="foo"
    full={true}
    title="Bar Title"
  />
</styled.div>
`;
